File: dsa_leetcode_beginner/0572-subtree-of-another-tree/0572-subtree-of-another-tree.cpp

#include <iostream>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (subRoot == nullptr) {
            return true;
        }
        if (root == nullptr) {
            return false;
        }
        if (isSameTree(root, subRoot)) {
            return true;
        }
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }

private:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr && q == nullptr) {
            return true;
        }
        if (p == nullptr || q == nullptr || p->val != q->val) {
            return false;
        }
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

// Example usage:
// TreeNode* root = new TreeNode(1);
// TreeNode* subRoot = new TreeNode(1);
// Solution solution;
// std::cout << solution.isSubtree(root, subRoot) << std::endl;
        

File: dsa_leetcode_beginner/0572-subtree-of-another-tree/0572-subtree-of-another-tree.js

class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Solution {
    isSubtree(root, subRoot) {
        if (subRoot === null) {
            return true;
        }
        if (root === null) {
            return false;
        }
        if (this.isSameTree(root, subRoot)) {
            return true;
        }
        return this.isSubtree(root.left, subRoot) || this.isSubtree(root.right, subRoot);
    }

    isSameTree(p, q) {
        if (p === null && q === null) {
            return true;
        }
        if (p === null || q === null || p.val !== q.val) {
            return false;
        }
        return this.isSameTree(p.left, q.left) && this.isSameTree(p.right, q.right);
    }
}

// Example usage:
// const root = new TreeNode(1);
// const subRoot = new TreeNode(1);
// const solution = new Solution();
// console.log(solution.isSubtree(root, subRoot));
        